### PR TITLE
feat: restore wallet from keystore JSON file 

### LINKED ISSUE #2661: [Demand]  Restore wallet from keystore JSON file
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

Restore the Ethereum wallet from a JSON key store file.

### Steps

+ Step 1: Read a JSON file.
+ Step 2: Validate the content with key store JSON schema. If failed then throw an error.
+ Step 3: Use lib to restore private key from the key store JSON.
+ Step 4: Recover wallet from the private key and a user given password.

### UI

+ Add a new tab - **Keystore** in the `<DashboardWalletImportDialog />`.
+ A multi-line text field or drop down area for the JSON file.
+ A single-line text field for the password.

### Screenshots

image
image

## Notes

+ [keythereum](https://github.com/ethereumjs/keythereum)
+ [ethereumjs/ethereumjs-wallet](https://github.com/ethereumjs/ethereumjs-wallet)

### PR DESCRIPTION
**Pull request by lelenei**
closes #2661

### CODE REVIEW

File: packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            Buffer.from(kdfparams.salt, 'hex'),
+            kdfparams.c,
+            kdfparams.dklen,
+            'sha256',
+        )
+    } else {
+        throw new Error('Unsupport key derivation scheme')
+    }
+    const ciphertext = Buffer.from(json.crypto.ciphertext, 'hex')
+    const mac = keccak256(Buffer.concat([Buffer.from(derivedKey.slice(16, 32)), ciphertext]))
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts:61]: Use `Web3Utils.keccak256`?

File: packages/maskbook/src/extension/options-page/DashboardDialogs/Wallet.tsx [hunk 2]
```diff
@@ -373,22 +446,23 @@ export function DashboardWalletImportDialog(props: WrappedDialogProps<object>) {
         props.onClose()
         onCreate(name)
     }, [state[0], name, hdWallet?.address, onCreate, onDeriveOrImport])
-
+    // zhuwp
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/Wallet.tsx:449]: What's this?

File: packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+export function exportFromV3Keystore(input: string | V3Keystore, password: string) {
+    const json: V3Keystore = typeof input === 'object' ? input : JSON.parse(input)
+
+    if (json.version !== 3) {
+        throw new Error('Not a V3 wallet')
+    }
+
+    let derivedKey: Uint8Array, kdfparams: any
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts:34]: Why any?

File: packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            salt: string
+        }
+        mac: string
+    }
+    id: string
+    version: number
+    address: string
+}
+
+export function exportFromV3Keystore(input: string | V3Keystore, password: string) {
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts:27]: Make it as a service function.

File: packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts (hunk 3 above)
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts:34]: Why any?

File: packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts [hunk 5]
```diff
@@ -58,9 +67,9 @@ export function exportFromV3Keystore(input: string | V3Keystore, password: strin
         throw new Error('Unsupport key derivation scheme')
     }
     const ciphertext = Buffer.from(json.crypto.ciphertext, 'hex')
-    const mac = keccak256(Buffer.concat([Buffer.from(derivedKey.slice(16, 32)), ciphertext]))
-    console.log(mac.toString('hex'))
-    if (mac.toString('hex') !== json.crypto.mac) {
+    const buf = Buffer.concat([Buffer.from(derivedKey.slice(16, 32)), ciphertext])
+    const mac = Web3Utils.sha3(`0x${buf.toString('hex')}`)
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts:71]: @guanbinrui is removing web3 from the repo. I don't think add new usage to web3 is a good idea

File: packages/maskbook/src/extension/options-page/DashboardDialogs/Wallet.tsx [hunk 6]
```diff
@@ -404,7 +404,7 @@ export function DashboardWalletImportDialog(props: WrappedDialogProps<object>) {
                             _private_key_: privateKey,
                         })
                     } catch (error) {
-                        throw new Error(error)
+                        throw error
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/Wallet.tsx:407]: so why you need a try-catch?

**Review by Jack-Works — CHANGES_REQUESTED** [review]
cc @guanbinrui

File: packages/maskbook/src/_locales/en/messages.json [hunk 7]
```diff
@@ -512,5 +512,13 @@
     "wallet_transfer_error_amount_absence": "Enter an amount",
     "wallet_transfer_error_address_absence": "Enter recipient address",
     "wallet_transfer_error_invalid_address": "Invalid recipient address",
-    "wallet_transfer_error_insufficent_balance": "Insufficient {{symbol}} balance"
+    "wallet_transfer_error_insufficent_balance": "Insufficient {{symbol}} balance",
+    "wallet_import_json": "Json Data",
```
Comment (septs) [inline packages/maskbook/src/_locales/en/messages.json:516]:
```suggestion
    "wallet_import_json": "JSON Data",
```

File: packages/maskbook/src/_locales/en/messages.json [hunk 8]
```diff
@@ -512,5 +512,13 @@
...
+    "wallet_import_json_keystore": "Keystore",
+    "wallet_import_json_keystore_content": "Paste keystore content",
+    "wallet_import_json_keystore_password": "Keystore Password",
+    "wallet_import_json_keystore_password_helper": "Enter keystore password",
+    "wallet_import_json_wallet_name": "Wallet Name",
+    "wallet_import_json_wallet_name_helper": "Set wallet name",
+    "wlalet_import_json_confirmation_hint": "I Have read and agree with the Tos & Privacy..."
```
Comment (septs) [inline packages/maskbook/src/_locales/en/messages.json:523]:
```suggestion
    "wlalet_import_json_confirmation_hint": "I have read and agree with the ToS & Privacy..."
```

File: packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts (hunk 5 above)
Comment (guanbinrui) [reply packages/maskbook/src/extension/options-page/DashboardDialogs/keystore.ts:71]: It's OK for this PR. I will replace it with another utility function.

File: packages/maskbook/src/_locales/en/messages.json (hunk 8 above)
Comment (guanbinrui) [reply packages/maskbook/src/_locales/en/messages.json:523]: So, where is the ToS doc?

File: packages/maskbook/src/extension/options-page/DashboardDialogs/Wallet.tsx [hunk 9]
```diff
@@ -74,7 +74,6 @@ import { useERC721TokenDetailed } from '../../../web3/hooks/useERC721TokenDetail
 import { useERC721TokenAssetDetailed } from '../../../web3/hooks/useERC721TokenAssetDetailed'
 import { CollectibleContext } from '../DashboardComponents/CollectibleList'
 
-//#region predefined token selector
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/Wallet.tsx:77]: Why delete this without removing the `//#endregion`?

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
+import scrypt from 'scrypt-js'
+import * as crypto from 'crypto'
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:2]: Please DO NOT use the `crypto` comes from the NodeJS std. Use Web Crypto API instead

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
+import scrypt from 'scrypt-js'
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:1]: Please use https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API.

**Review by Jack-Works — CHANGES_REQUESTED** [review]

**Comment by nuanyang233** [comment]
plz resolve confilct

**Review by guanbinrui — DISMISSED** [review]

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts (hunk 11 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/Wallet/services/keystore.ts:1]: this is not resolved.

File: packages/maskbook/package.json [hunk 12]
```diff
@@ -94,6 +94,7 @@
     "react-virtualized-auto-sizer": "^1.0.2",
     "react-window": "^1.8.6",
     "remarkable": "^2.0.1",
+    "scrypt-js": "^3.0.1",
```
Comment (Jack-Works) [inline packages/maskbook/package.json:97]: Please use Web Crypto API

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+    interface PBKDF2Params {
+        c: number
+        prf: 'hmac-sha256'
+        dklen: number
+        salt: string
+    }
+
+    interface Scrypt {
+        kdf: 'scrypt'
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:34]: Oh, so this is a new kind of kdf algr we're supporting? cc @yisiliu

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+        'aes-128-ctr'
+            ? {
+                  name: 'AES-CTR',
+                  counter: iv,
+                  length: 128,
+              }
+            : {
+                  name: 'AES-CBC',
+                  iv,
+              },
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:138]: what's the meaning of this part? it always the first branch matches

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        key,
+        ciphertext,
+    )
+    return Buffer.from(seed).toString('hex')
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:142]: Please explicit import `buffer` module

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/package.json (hunk 12 above)
Comment (lelenei) [reply packages/maskbook/package.json:97]: web crypto api don't support scrypt

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts (hunk 11 above)
Comment (lelenei) [reply packages/maskbook/src/plugins/Wallet/services/keystore.ts:1]: web crypto don't support scrypt

**Review by guanbinrui — DISMISSED** [review]

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+export interface V3Keystore {
+    crypto: KeyStore
+    id: string
+    version: number
+    address: string
+}
+
+export async function fromV3Keystore(input: string | V3Keystore, password: string) {
+    const json: V3Keystore = typeof input === 'object' ? input : (JSON.parse(input) as V3Keystore)
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:56]: use json-schema, verify data structure.

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+
+    if (json.version !== 3) {
+        throw new Error('Not a V3 wallet')
+    }
+
+    let derivedKey: Uint8Array, kdfparams: KeyStore.ScryptParams | KeyStore.PBKDF2Params
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:62]:
```suggestion
    let derivedKey: Uint8Array
    let kdfparams: KeyStore.ScryptParams | KeyStore.PBKDF2Params
```

split variable definitions.

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+        Buffer.from(json.crypto.ciphertext, 'hex'),
+        Buffer.from(json.crypto.cipherparams.iv, 'hex'),
+    )
+    return { address: `0x${json.address}`, privateKey: `0x${seed}` } as {
+        address: string
+        privateKey: string
+    }
+}
+
+async function Decrypt(cipher: string, derivedKey: Uint8Array, ciphertext: Uint8Array, iv: Uint8Array) {
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:117]:
```suggestion
async function decrypt(cipher: string, derivedKey: Uint8Array, ciphertext: Uint8Array, iv: Uint8Array) {
```

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+        throw new Error('Unsupport key derivation scheme')
+    }
+    const ciphertext = Buffer.from(json.crypto.ciphertext, 'hex')
+    const buf = Buffer.concat([Buffer.from(derivedKey.slice(16, 32)), ciphertext])
+    const mac = Web3Utils.sha3(`0x${buf.toString('hex')}`)
+    if (mac !== `0x${json.crypto.mac}`) {
+        throw new Error('Key derivation failed - possibly wrong passphrase')
+    }
+
+    const seed = await Decrypt(
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:105]:
```suggestion
    const seed = await decrypt(
```

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts (hunk 16 above)
Comment (lelenei) [reply packages/maskbook/src/plugins/Wallet/services/keystore.ts:56]: the json schema you provided don't seem to be verified.

Comment (guanbinrui) [reply packages/maskbook/src/plugins/Wallet/services/keystore.ts:56]: It's ok to leave a TODO for now.

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 20]
```diff
@@ -59,8 +59,10 @@ export async function fromV3Keystore(input: string | V3Keystore, password: strin
         throw new Error('Not a V3 wallet')
     }
 
+    // TODO; check json schema
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:62]:
```suggestion
    // TODO: use json schema check input
    // see https://github.com/DimensionDev/Maskbook/issues/2661
```

**Review by guanbinrui — DISMISSED** [review]

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+
+export interface V3Keystore {
+    crypto: KeyStore
+    id: string
+    version: number
+    address: string
+}
+
+export async function fromV3Keystore(input: string | V3Keystore, password: string) {
+    const json: V3Keystore = typeof input === 'object' ? input : (JSON.parse(input) as V3Keystore)
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:56]:
```suggestion
    const json: V3Keystore = typeof input === 'object' ? input : JSON.parse(input)
```

**Review by Jack-Works — DISMISSED** [review]
Might be OK but need test

**Comment by guanbinrui** [comment]
image

@lelenei Please change the button state to loading (present a spinner) while restoring the Keystore file.

**Comment by lelenei** [comment]
> image
> 
> @lelenei Please change the button state to loading (present a spinner) while restoring the Keystore file.
> 
> P.s. Also, I've realized that the UI is blocking while restoring the Keystore file.
> P.s. Cannot restore the wallet with the private key on this PR.

Post the contents of the keystore file in this PR.

File: packages/maskbook/src/plugins/Wallet/services/keystore.schema.json [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+                        "salt": { "$ref": "#/definitions/encoded-string" }
+                    },
+                    "required": ["c", "prf", "dklen", "salt"],
+                    "additionalProperties": false
+                }
+            }
+        },
+        "scrypt": {
+            "type": "object",
+
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.schema.json:75]: plz remove the empty line.

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 23]
```diff
@@ -1,6 +1,8 @@
 import scrypt from 'scrypt-js'
 import Web3Utils from 'web3-utils'
 import { Buffer } from 'buffer'
+import keystoreSchema from './keystore.schema.json'
+import z_schema from 'z-schema'
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:5]: ??? @Jack-Works

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 24]
```diff
@@ -52,15 +54,27 @@ export interface V3Keystore {
     address: string
 }
 
-export async function fromV3Keystore(input: string | V3Keystore, password: string) {
-    const json: V3Keystore = typeof input === 'object' ? input : JSON.parse(input)
+function ValidatKeystore(input: string | V3Keystore) {
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:57]: typo error, `validate`

File: packages/maskbook/src/extension/options-page/DashboardComponents/ActionButton.tsx [hunk 25]
```diff
@@ -58,8 +58,9 @@ export function DebounceButton(_props: DebounceButtonProps) {
             onClick={f}
             classes={color === 'danger' ? classes : undefined}
             color={color === 'danger' ? 'primary' : color}
-            {...props}
-        />
+            {...props}>
+            {loading ? 'loading...' : _props.children}
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardComponents/ActionButton.tsx:62]:
```suggestion
            {loading ? null : _props.children}
```

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts (hunk 23 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/Wallet/services/keystore.ts:5]: We're using z-schema to verify if an JSON object satisfy a JSON Schema

**Comment by Jack-Works** [comment]
hi @lelenei please rebase/merge with master

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 26]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+export interface V3Keystore {
+    crypto: KeyStore
+    id: string
+    version: number
+    address: string
+}
+
+function ValidateKeystore(input: string | V3Keystore) {
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:57]:
```suggestion
function ValidateKeystore(input: string) {
```

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts [hunk 27]
```diff
@@ -0,0 +1,161 @@
...
+        const valid = validator.validate(json, keystoreSchema)
+        if (!valid) return null
+        return json
+    } catch {
+        return null
+    }
+}
+
+export async function fromV3Keystore(input: string | V3Keystore, password: string) {
+    const json = ValidateKeystore(input)
```
Comment (septs) [inline packages/maskbook/src/plugins/Wallet/services/keystore.ts:70]:
```suggestion
export async function fromV3Keystore(input: string, password: string) {
    const keyStore = ValidateKeyStore(JSON.parse(input))
```

File: packages/maskbook/src/plugins/Wallet/services/keystore.ts (hunk 26 above)
Comment (septs) [reply packages/maskbook/src/plugins/Wallet/services/keystore.ts:57]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

**Comment by guanbinrui** [comment]
Replaced by #3404

**Pull request closed without merging**